codex/slots/proofs/backends/circomcompat.nim

```diff
@@ -0,0 +1,237 @@
+  # to the correct length
+  doAssert input.samples.len == self.numSamples,
+    "Number of samples does not match"
+
+  doAssert input.slotProof.len <= self.datasetDepth,
+    "Number of slot proofs does not match"
+
+  doAssert input.samples.allIt(
+    block:
+      (it.merklePaths.len <= self.slotDepth + self.blkDepth and
+      it.cellData.len <= self.cellElms * 32)), "Merkle paths length does not match"
+
+  # TODO: All parameters should match circom's static parametter
+  var
+    backend: ptr CircomCompatCtx
+
+  if initCircomCompat(
+    self.backendCfg,
+    addr backend) != ERR_OK or backend == nil:
+    raiseAssert("failed to initialize CircomCompat backend")
+
+  var
+    entropy = input.entropy.toBytes
+    dataSetRoot = input.datasetRoot.toBytes
+    slotRoot = input.slotRoot.toBytes
+
+  if backend.pushInputU256Array(
+    "entropy".cstring, entropy[0].addr, entropy.len.uint32) != ERR_OK:
+    return failure("Failed to push entropy")
+
+  if backend.pushInputU256Array(
+    "dataSetRoot".cstring, dataSetRoot[0].addr, dataSetRoot.len.uint32) != ERR_OK:
+    return failure("Failed to push data set root")
+
+  if backend.pushInputU256Array(
+    "slotRoot".cstring, slotRoot[0].addr, slotRoot.len.uint32) != ERR_OK:
+    return failure("Failed to push data set root")
+
+  if backend.pushInputU32(
+    "nCellsPerSlot".cstring, input.nCellsPerSlot.uint32) != ERR_OK:
+    return failure("Failed to push nCellsPerSlot")
+
+  if backend.pushInputU32(
+    "nSlotsPerDataSet".cstring, input.nSlotsPerDataSet.uint32) != ERR_OK:
+    return failure("Failed to push nSlotsPerDataSet")
+
+  if backend.pushInputU32(
+    "slotIndex".cstring, input.slotIndex.uint32) != ERR_OK:
+    return failure("Failed to push slotIndex")
+
+  var
+    slotProof = input.slotProof.mapIt( it.toBytes ).concat
+
+  slotProof.setLen(self.datasetDepth) # zero pad inputs to correct size
+
+  # arrays are always flattened
+  if backend.pushInputU256Array(
+    "slotProof".cstring,
+    slotProof[0].addr,
+    uint (slotProof[0].len * slotProof.len)) != ERR_OK:
+      return failure("Failed to push slot proof")
+
+  for s in input.samples:
+    var
+      merklePaths = s.merklePaths.mapIt( it.toBytes )
+      data = s.cellData
+
+    merklePaths.setLen(self.slotDepth) # zero pad inputs to correct size
+    if backend.pushInputU256Array(
+      "merklePaths".cstring,
+      merklePaths[0].addr,
+      uint (merklePaths[0].len * merklePaths.len)) != ERR_OK:
+        return failure("Failed to push merkle paths")
+
+    data.setLen(self.cellElms * 32) # zero pad inputs to correct size
+    if backend.pushInputU256Array(
+      "cellData".cstring,
+      data[0].addr,
+      data.len.uint) != ERR_OK:
+        return failure("Failed to push cell data")
+
+  var
+    proofPtr: ptr Proof = nil
+
+  let proof =
+    try:
+      if self.backendCfg.proveCircuit(backend, proofPtr.addr) != ERR_OK or
+        proofPtr == nil:
+        return failure("Failed to prove")
```

Comment:
Yuck, we get zero error info back from the backend on failure? :-/

Reply:
we get an error code, perhaps we should log it